src/bio/bio_wal.c

```diff
@@ -0,0 +1,1191 @@
+	     unsigned int hdr_sz, uint32_t *csum)
+{
+	bio_addr_t	addr = { 0 };
+	d_iov_t		iov;
+	int		rc, csum_len;
+
+	csum_len = meta_csum_len(mc);
+	rc = meta_csum_calc(mc, hdr, hdr_sz - csum_len, csum, csum_len);
+	if (rc) {
+		D_ERROR("Calculate headr csum failed. "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	bio_addr_set(&addr, DAOS_MEDIA_NVME, 0);
+	d_iov_set(&iov, hdr, hdr_sz);
+
+	rc = bio_write(ioc, addr, &iov);
+	if (rc) {
+		D_ERROR("Failed to write header. "DF_RC"\n", DP_RC(rc));
+		return rc;
+	}
+
+	return 0;
+}
+
+static int
+flush_wal_header(struct bio_meta_context *mc)
+{
+	struct wal_super_info	*si = &mc->mc_wal_info;
+	struct wal_header	*hdr = &si->si_header;
+
+	/* WAL header is up-to-date */
+	if (si->si_ckp_id == hdr->wh_ckp_id && si->si_ckp_blks == hdr->wh_ckp_blks &&
+	    si->si_commit_id == hdr->wh_commit_id && si->si_commit_blks == hdr->wh_commit_blks)
+		return 0;
+
+	hdr->wh_ckp_id = si->si_ckp_id;
+	hdr->wh_ckp_blks = si->si_ckp_blks;
+	hdr->wh_commit_id = si->si_commit_id;
+	hdr->wh_commit_blks = si->si_commit_blks;
+
+	return write_header(mc, mc->mc_wal, hdr, sizeof(*hdr), &hdr->wh_csum);
+}
+
+int
+bio_wal_replay(struct bio_meta_context *mc,
+	       int (*replay_cb)(struct umem_action *actv, unsigned int act_nr),
+	       unsigned int max_replay_nr)
+{
+	/* TODO */
+	return 0;
+}
+
+int
+bio_wal_ckp_start(struct bio_meta_context *mc, uint64_t *tx_id)
+{
+	/* TODO */
+	return 0;
+}
+
+int
+bio_wal_ckp_end(struct bio_meta_context *mc, uint64_t tx_id)
+{
+	/* TODO */
+	return 0;
+}
+
+static void
+unload_wal(struct bio_meta_context *mc)
+{
+	struct wal_super_info	*si = &mc->mc_wal_info;
+	int			 rc;
+
+	D_ASSERT(d_list_empty(&si->si_pending_list));
+	D_ASSERT(si->si_tx_failed == 0);
+	if (si->si_rsrv_waiters > 0)
+		wakeup_reserve_waiters(si, true);
+
+	rc = flush_wal_header(mc);
+	if (rc)
+		D_ERROR("Flush WAL header failed. "DF_RC"\n", DP_RC(rc));
+
+	ABT_mutex_free(&si->si_mutex);
+	ABT_cond_free(&si->si_rsrv_wq);
+}
+
+static int
+load_wal(struct bio_meta_context *mc)
+{
+	struct wal_super_info	*si = &mc->mc_wal_info;
+	struct wal_header	*hdr = &si->si_header;
+	int			 rc;
+
+	rc = load_wal_header(mc);
+	if (rc)
+		return rc;
+
+	rc = ABT_mutex_create(&si->si_mutex);
+	if (rc != ABT_SUCCESS)
+		return -DER_NOMEM;
+
+	rc = ABT_cond_create(&si->si_rsrv_wq);
+	if (rc != ABT_SUCCESS) {
+		ABT_mutex_free(&si->si_mutex);
+		return -DER_NOMEM;
+	}
+
+	D_INIT_LIST_HEAD(&si->si_pending_list);
+	si->si_rsrv_waiters = 0;
+	si->si_tx_failed = 0;
+
+	si->si_ckp_id = hdr->wh_ckp_id;
+	si->si_ckp_blks = hdr->wh_ckp_blks;
+	si->si_commit_id = hdr->wh_commit_id;
+	si->si_commit_blks = hdr->wh_commit_blks;
+
+	D_ASSERTF(wal_id_cmp(si, si->si_ckp_id, si->si_commit_id) <= 0,
+		  "Checkpoint ID "DF_U64" > Committed ID "DF_U64"\n",
+		  si->si_ckp_id, si->si_commit_id);
+
+	si->si_unused_id = wal_next_id(si, si->si_commit_id, si->si_commit_blks);
+
+	return 0;
+
+}
+
+static int
+load_meta_header(struct bio_meta_context *mc)
```

Comment:
Could we export this function? considering following case:
 int bio_mc_open:
 |->__bio_ioctxt_open META
call load_meta_header() to check if this pool nvme size is 0 or not, and also
we need get WAL blob id for loading SMD case.

Reply:
Yes, sure.